Perform token validation against Adgangsplatformen

kasperg opened the pull request:
The OAuth library for Adgangsplatformen provides PSR-15 middleware
for validating access tokens and mapping these to resource owners.

The Laravel PSR-15 bridge provides the simplest way to use PSR-15
middleware in Lumen as it simply wraps an PSR-15 middleware object.
[The more popular alternative](https://github.com/jshannon63/laravel-psr15-middleware) is way too complex.

We use the middleware to map tokens to resource owners if available.
As the application currently expects a user in the string format we
use a subsequent middleware to map a resource owner to an id.

kasperg commented:
I thought about what the best way to provide both a setup production setup and a test which is not dependant on Adgangsplatformen.

The current approach seems to create a somewhat bigger gap between these two situations with more complex middlewares being used in the production setup. This also results in the decrease in coverage.

An alternative solution might be to change [the signature of the Middleware](https://github.com/reload/oauth2-adgangsplatformen/blob/master/src/Middleware/TokenResourceOwnerMapper.php#L23) to support injecting an `AbstractProvider`. We could then implement a dummy provider which returns dummy resource owners which use the authentication token as the id and use this provider for testing.

xendk commented on bootstrap/app.php:
```diff
@@ -46,6 +46,27 @@
     App\Console\Kernel::class
 );
 
+$app->singleton(
```
These shouldn't be here. We should have a service provider that registers the needed classes in the container. As they're not inside the conditional, I assume it'll work just as well in a service provider whether or not it's prod or test.

https://laravel.com/docs/5.8/providers

Rather than adapting `TokenResourceOwnerMapper` and then using adding another `ResourceOwnerGuidMapper` middleware, I was thinking of letting `TokenChecker` pass the request to `TokenResourceOwnerMapper` and copying over the GUID to the original request. 

One could provide a service provider in `oauth2-adgangsplatformen`, but that would add `illuminate/support` as a dependency. OTOH, one could leave out the dependency, and rely on users only using the AdgangsplatformenServiceProvider class from Laravel/Lumen projects.

kasperg commented:
I have updated the PR with some changes based on your suggestions:

1. I have added two providers for configuring Adgangsplatformen. One which is always used and one which overrides the OAuth provider for testing. This reduces logic in `app.php` and there is no longer need for `TestTokenChecker` which can then be removed.
2. I intentionally placed the providers in this repo for now. I think there is value in having `oauth2-adgangsplatformen` be decoupled from Laravel/Lumen as long as this is the only service using this framework and there are [other services](danskernesdigitalebibliotek/ddb-cover-service) using something else. If things progress and usage of Laravel/Lumen expand we could either add a service provider (and wrapped middleware) using [suggested dependencies](https://github.com/reload/oauth2-adgangsplatformen/blob/master/composer.json#L24-L29) or by creating a separate `laravel-adgangsplatformen` package. (If we created a separate Laravel-specific package perhaps the CLI app could be migrated to an artisan command as well 🤔 )
3. Instead of splitting the authorization logic over multiple different middlewares I think there is value in keeping it together. It allows us to reuse the logic across multiple services and frameworks and we can also produce better error messages. I made a suggestion for how this could be achieved in github.com/reload/oauth2-adgangsplatformen#2. If we take this route we can also drop the current `TokenAccess` middleware.
4. I agree that the `handle` method in `TokenResourceOwnerMapper` may be too complex because it tries to juggle multiple Authorization headers. Perhaps it should just be reduced to one. [That is what others do as well](https://github.com/thephpleague/oauth2-server/blob/master/src/AuthorizationValidators/BearerTokenValidator.php#L66).

xendk commented:
This still rubs me the wrong way.

Why are we insisting on using PSR-15? It's not in the spec (which only requires being based on Symfony), neither Laravel/Lumen nor Symfony is using it per default, and no other project we should support is using it. Why are we making our only use case more complicated?

To quote Kevlin Henney from "97 Things Every Software Architect Should Know":

> Favoring simplicity before generality acts as a tiebreaker between otherwise equally viable design alternatives. When there are two possible solutions, favor the one that is simpler and based on concrete need rather than the more intricate one that boasts of generality. Of course, it is entirely possible (and more than a little likely) that the simpler solution will turn out to be the more general one in practice. And if that doesn't turn out to be the case, it will be easier to change the simpler solution to what you now know you need than to change the 'general' one that turns out not to be quite general enough in the right way.

Reuse is the point of the Adgangsplatformen project, so why are we making it *harder* to reuse in Laravel/Lumen, which we suspect will be used by some of the other pipelined projects too? Why doesn't Adgangsplatformen have a middleware that'll set a `AdgangsplatformenUser` as the request user? It might even implement a [driver](https://culttt.com/2015/08/31/how-to-resolve-environment-specific-implementations-from-laravels-ioc-container/) that allows for configuring a testing implementation. Adgangsplatformen doesn't need to depend on illuminate, and can still contain a PSR-15 implementation if someone wants it.

kasperg commented:
@xendk I reworked the PR based on our discussion.

First of all please check reload/oauth2-adgangsplatformen#3 which this PR depends upon.

Key changes are:

1. All necessary providers and middleware are added to the application
by the provider. The bootstrap becomes much easier.
2. The token validation middleware is registered using the auth
shorthand. Consequently we add all routes requiring authentication
to this. This also means you can check the version of the service
without authentication.
3. The middleware now ensures that a domain specific user
object is returned for authenticated requests and not just a string.
Consequently we have to call the proper method to actually get the
user id in our controller.
4. The middleware is also responsible for rejecting unauthenticated
requests so we can drop the TokenAccess middleware entirely.
5. There is no need to define which driver Adgangsplatformen uses. For
testing it will automatically use the testing version.

Note that this currently depends on a branch of the Adgangsplatformen
library. This will be reverted to master once the changes there are
merged.

kasperg commented:
> why are you changing the fake ids in the tests?

I tried to explain in the commit comment for 2620dcd56110a0f17c3c72b1d28f1b035173a8ef.
